File: src/commands/remove.ts
import * as p from '@clack/prompts';
import chalk from 'chalk';
import { join } from 'node:path';
import { exists } from '../utils/fs.js';
import { isValidAddon, uninstallAddon } from '../addons/index.js';

export async function remove(addon: string): Promise<void> {
	const cwd = process.cwd();

	p.intro(chalk.bold(`forge remove ${addon}`));

	if (!isValidAddon(addon)) {
		p.log.error(`Unknown addon: "${addon}"`);
		process.exit(1);
	}

	if (!(await exists(join(cwd, 'forge.yaml')))) {
		p.log.error('No forge.yaml found. Run `forge init` first.');
		process.exit(1);
	}

	// Check if installed
	const { readYaml } = await import('../utils/yaml.js');
	const config = await readYaml<Record<string, unknown>>(join(cwd, 'forge.yaml'));
	const addons = (config.addons as string[]) ?? [];
	if (!addons.includes(addon)) {
		p.log.warn(`Addon "${addon}" is not installed.`);
		process.exit(0);
	}

	const spinner = p.spinner();
	spinner.start(`Removing ${addon}...`);

	try {
		const removedFiles = await uninstallAddon(addon, cwd);

		for (const file of removedFiles) {
			p.log.success(`Deleted ${file}`);
		}
		p.log.success('Updated forge.yaml');

		spinner.stop(`${addon} removed`);
	} catch (err) {
		spinner.stop('Removal failed');
		p.log.error(String(err));
		process.exit(1);
	}

	p.outro(chalk.green('Done!'));
}

File: cmd/remove.go
package cmd

import (
	"fmt"
	"github.com/samahlstrom/forge-cli/internal/addon"
	"github.com/samahlstrom/forge-cli/internal/ui"
	"github.com/samahlstrom/forge-cli/internal/util"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{
		Use:   "remove <addon>",
		Short: "Remove an installed addon",
		Args:  cobra.ExactArgs(1),
		RunE:  runRemove,
	})
}

func runRemove(cmd *cobra.Command, args []string) error {
	name := args[0]
	cwd, _ := os.Getwd()

	ui.Intro(ui.Bold(fmt.Sprintf("forge remove %s", name)))

	if !addon.IsValid(name) {
		ui.Log.Error(fmt.Sprintf("Unknown addon: %q", name))
		os.Exit(1)
	}

	if !util.Exists(filepath.Join(cwd, "forge.yaml")) {
		ui.Log.Error("No forge.yaml found. Run `forge init` first.")
		os.Exit(1)
	}

	var config map[string]any
	_ = util.ReadYAML(filepath.Join(cwd, "forge.yaml"), &config)
	addons := toStringSlice(config["addons"])
	found := false
	for _, a := range addons {
		if a == name {
			found = true
			break
		}
	}
	if !found {
		ui.Log.Warn(fmt.Sprintf("Addon %q is not installed.", name))
		os.Exit(0)
	}

	spinner := ui.NewSpinner()
	spinner.Start(fmt.Sprintf("Removing %s...", name))

	files, err := addon.Uninstall(name, cwd)
	if err != nil {
		spinner.Stop("Removal failed")
		ui.Log.Error(err.Error())
		os.Exit(1)
	}

	for _, f := range files {
		ui.Log.Success(fmt.Sprintf("Deleted %s", f))
	}
	ui.Log.Success("Updated forge.yaml")

	spinner.Stop(fmt.Sprintf("%s removed", name))
	ui.Outro(ui.Green("Done!"))
	return nil
}
